clustered/server/src/main/java/org/ehcache/clustered/server/internal/messages/EhcacheSyncMessageCodec.java

```diff
@@ -101,26 +107,33 @@
   public byte[] encode(final int concurrencyKey, final EhcacheEntityMessage message) throws MessageCodecException {
     if (message instanceof EhcacheSyncMessage) {
       EhcacheSyncMessage syncMessage = (EhcacheSyncMessage) message;
-      StructEncoder encoder;
+      StructEncoder<Void> encoder;
       switch (syncMessage.getMessageType()) {
-        case STATE:
+        case STATE: {
           encoder = STATE_SYNC_STRUCT.encoder();
-          EhcacheStateSyncMessage stateSyncMessage = (EhcacheStateSyncMessage) syncMessage;
+          EhcacheStateSyncMessage stateSyncMessage = (EhcacheStateSyncMessage)syncMessage;
           encoder.enm(SYNC_MESSAGE_TYPE_FIELD_NAME, STATE);
           encodeServerSideConfiguration(encoder, stateSyncMessage.getConfiguration());
           encoder.structs(STORES_SUB_STRUCT, stateSyncMessage.getStoreConfigs().entrySet(), (storeEncoder, storeEntry) -> {
             storeEncoder.string(SERVER_STORE_NAME_FIELD, storeEntry.getKey());
             codecUtils.encodeServerStoreConfiguration(storeEncoder, storeEntry.getValue());
           });
           return encoder.encode().array();
-        case DATA:
+        }
+        case DATA: {
           encoder = DATA_SYNC_STRUCT.encoder();
-          EhcacheDataSyncMessage dataSyncMessage = (EhcacheDataSyncMessage) syncMessage;
+          EhcacheDataSyncMessage dataSyncMessage = (EhcacheDataSyncMessage)syncMessage;
           encoder.enm(SYNC_MESSAGE_TYPE_FIELD_NAME, DATA);
           encoder.string(SERVER_STORE_NAME_FIELD, dataSyncMessage.getCacheId());
-          encoder.int64(KEY_FIELD, dataSyncMessage.getKey());
-          encoder.struct(CHAIN_FIELD, (chainEncoder) -> chainCodec.encode(chainEncoder, dataSyncMessage.getChain()));
+          StructArrayEncoder entriesEncoder = encoder.structs(CHAIN_MAP_ENTRIES_SUB_STRUCT);
+          for (Map.Entry<Long, Chain> entry : dataSyncMessage.getChainMap().entrySet()) {
+            entriesEncoder.int64(KEY_FIELD, entry.getKey());
+            chainCodec.encode(entriesEncoder.struct(CHAIN_FIELD), entry.getValue());
+            entriesEncoder.next();
+          }
+          entriesEncoder.end();
```

Comment:
You could replace this loop with a call to `encoder.structs(name, lambda)` after setting `StructArrayEncoder`'s generic if you wanted to.